distributed/semaphore.py

```diff
@@ -0,0 +1,327 @@
+            elapsed = time() - self.started_at
+            return max(0, self.duration - elapsed)
+
+
+class SemaphoreExtension:
+    """ An extension for the scheduler to manage Semaphores
+
+    This adds the following routes to the scheduler
+
+    * semaphore_acquire
+    * semaphore_release
+    * semaphore_create
+    """
+
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.leases = defaultdict(deque)
+        self.events = defaultdict(asyncio.Event)
+        self.max_leases = dict()
+        self.leases_per_client = defaultdict(partial(defaultdict, deque))
+        self.scheduler.handlers.update(
+            {
+                "semaphore_create": self.create,
+                "semaphore_acquire": self.acquire,
+                "semaphore_release": self.release,
+                "semaphore_close": self.close,
+            }
+        )
+
+        self.scheduler.extensions["semaphores"] = self
+        self.pc_validate_leases = PeriodicCallback(
+            self._validate_leases,
+            1000
+            * parse_timedelta(
+                dask.config.get(
+                    "distributed.scheduler.locks.lease-validation-interval"
+                ),
+                default="s",
+            ),
+            io_loop=self.scheduler.loop,
+        )
+        self.pc_validate_leases.start()
+        self._validation_running = False
+
+    # `comm` here is required by the handler interface
+    async def create(self, comm=None, name=None, max_leases=None):
+        # We use `self.max_leases.keys()` as the point of truth to find out if a semaphore with a specific
+        # `name` has been created.
+        if name not in self.max_leases:
+            assert isinstance(max_leases, int), max_leases
+            self.max_leases[name] = max_leases
+        else:
+            if max_leases != self.max_leases[name]:
+                raise ValueError(
+                    "Inconsistent max leases: %s, expected: %s"
+                    % (max_leases, self.max_leases[name])
+                )
+
+    async def _get_lease(self, client, name, identifier):
+        # We should make sure that the client is already properly registered with the scheduler
+        # otherwise the lease validation will mop up every acquired lease immediately. That's mostly relevant for tests
+        while client not in self.scheduler.clients:
+            await asyncio.sleep(0.0005)  # This value is set somewhat arbitrarily
+
+        result = True
+        if len(self.leases[name]) < self.max_leases[name]:
+            # naive: self.leases[resource] += 1
+            # not naive:
+            self.leases[name].append(identifier)
+            self.leases_per_client[client][name].append(identifier)
+        else:
+            result = False
+        return result
+
+    def _semaphore_exists(self, name):
+        if name not in self.max_leases:
+            logger.exception(f"Semaphore `{name}` not known or already closed.")
```

Comment:
@mrocklin how about the handling of this error? Context: the semaphore has been closed using `Semaphore.close` and a client attempts to acquire/release/close the closed semaphore. Would the expected behavior be that there is an exception raised?

Reply:
I would not log or raise here since the behaviour should be depending on where this check is performed. I think it's ok to encapsulate the _exists_ check in a method since what data structure needs to be checked is not trivial

Reply:
I'm in agreement with @fjetter on this one. The post condition is met and the semaphore released. No need to raise an exception.

Reply:
To make this concrete from the Client perspective, is this the situation that we're considering?

```python
# client 1
semaphore = Semaphore(name="my-resource")
# do things with semaphore
semaphore.close()
```

```python
# client 2
semaphore = Semaphore(name="my-resource")
ValueError("Semaphore already closed")
```

If so I agree that this shouldn't occur.  Perhaps this isn't what we're discussing though.  If folks here are able to spell out what the result of this code is that would help in review (sorry, there is a large backlog to go through, so I'm rushing through reviews today)

Reply:
changed this behavior in https://github.com/dask/distributed/pull/3573/commits/7a519279a2539203f5dd2f18c0506365b6f28e4e

@mrocklin no, it is for the case when any client closes a semaphore and then any client tries to acquire, release or close that semaphore. See the commit I linked for the behavior for each of these operations now. 

Happy to edit that commit one final time if we can get to a consensus